OpenXR: Add support for spatial entities extension

BastiaanOlij opened the pull request:
[The OpenXR Spatial entities extension](https://www.khronos.org/blog/openxr-spatial-entities-extensions-released-for-developer-feedback) was introduced to standardise obtaining and interacting with information about the users real world environment.

It defines [a core extension](https://registry.khronos.org/OpenXR/specs/1.1/html/xrspec.html#XR_EXT_spatial_entity) for how to query and interact with spatial entities, and defines additional extensions that detail out specific uses of these entities.

Currently the core specification has support for:
- [Spatial anchors](https://registry.khronos.org/OpenXR/specs/1.1/html/xrspec.html#XR_EXT_spatial_anchor), this supports marking locations in the real world that can be made persistent. Ideal for placing virtual objects into the real world and having them reconstructed in future sessions
- [Plane tracking](https://registry.khronos.org/OpenXR/specs/1.1/html/xrspec.html#XR_EXT_spatial_plane_tracking), identifies various surfaces such as floors, walls, tables, etc. This allows us to know where surfaces are and provides the ability to place/attach virtual objects on/to these surfaces or apply physics to these surfaces.
- [Marker tracking](https://registry.khronos.org/OpenXR/specs/1.1/html/xrspec.html#XR_EXT_spatial_marker_tracking), visually detects markers in the real world and brings their position into the virtual world. Currently supported are QR codes, Aruca markers or April tags.

The spatial entities system is designed to be incredibly modular with future plans for additional (vendor) extensions to unlock further functionality. The core of the implementation in Godot therefor exposes a fair amount of underlying features that can be used in GDExtension plugins to implement such extensions or for users to fully customize the setup.

However, in plain vanila mode all this logic comes built in and can be enabled in project settings. Using the built in logic all spatial entities are recorded as Anchor trackers with the XR Server and can be consumed through a simple manager script.

While the implementation is deemed complete based on the published specification, runtime implementations are still being worked upon and some tweaking will be expected.

Documentation can be found here: https://github.com/godotengine/godot-docs/pull/11015
Demo project can be found here: https://github.com/BastiaanOlij/spatial-entities-demo

For testing sadly there are no public runtime implementations yet, however myself, David and Fredia have access to betas and have tested this.

*Contributed by Khronos Group through the Godot Integration Project*

BastiaanOlij committed:
OpenXR: Implement spatial entities extension

BastiaanOlij commented:
Note: Seeing we're pretty much in feature freeze for Godot 4.5, this PR is currently set for inclusion in Godot 4.6.

However if consensus is met in time, I'm hoping we can merge this as an experimental feature in Godot 4.5 as the functionality should not impact any users that do not enable it.

This is worth further discussion.

dsnopek reviewed (commented):
Thanks, this is great! I wish I could test it :-)

I skimmed through the code (I didn't look at the docs), and only have a few minor comments

BastiaanOlij commented on modules/openxr/extensions/spatial_entities/openxr_spatial_entities.cpp:
```diff
@@ -0,0 +1,494 @@
+Ref<Mesh> OpenXRSpatialComponentMesh3DList::get_mesh(int64_t p_index) const {
+	ERR_FAIL_INDEX_V(p_index, mesh3d_data.size(), nullptr);
+
+	// TODO implement, need to convert mesh data to Godot mesh resource
```
Oh yeah, I forgot about that one.

BastiaanOlij commented on modules/openxr/extensions/spatial_entities/openxr_spatial_entity_extension.cpp:
```diff
@@ -0,0 +1,1177 @@
+
+	ADD_SIGNAL(MethodInfo("spatial_discovery_recommended", PropertyInfo(Variant::RID, "spatial_context")));
+
+	// TODO component_types should be an int array typed to OpenXrSpatialComponentType, but we currently don't support that (I think).
```
Sads :)

BastiaanOlij commented on modules/openxr/extensions/spatial_entities/openxr_spatial_entity_extension.cpp:
```diff
@@ -0,0 +1,1177 @@
+	ERR_FAIL_NULL_V(openxr_api, RID());
+
+	// Convert our entity RIDs to XrSpatialEntityEXT
+	Vector<XrSpatialEntityEXT> entities;
```
Still a little fuzzy on when LocalVector is appropriate. We really need a good writeup of that some day :)

m4gr3d commented on modules/openxr/doc_classes/OpenXRSpatialAnchorCapability.xml:
```diff
@@ -0,0 +1,120 @@
+			<param index="1" name="persistence_context" type="RID" default="RID()" />
+			<param index="2" name="callback" type="Callable" default="Callable()" />
+			<description>
+				Remotes the persistent data from this anchor. The runtime will not recreate the anchor when your application restarts. If [param persistence_context] is not specified the default will be used, this requires [member ProjectSettings.xr/openxr/extensions/spatial_entity/enabled] to be set.
```
```suggestion
				Removes the persistent data from this anchor. The runtime will not recreate the anchor when your application restarts. If [param persistence_context] is not specified the default will be used, this requires [member ProjectSettings.xr/openxr/extensions/spatial_entity/enabled] to be set.
```

m4gr3d commented on modules/openxr/doc_classes/OpenXRSpatialAnchorCapability.xml:
```diff
@@ -0,0 +1,120 @@
+			<return type="void" />
+			<param index="0" name="anchor_tracker" type="OpenXRAnchorTracker" />
+			<description>
+				Remove an anchor previously created with [method create_new_anchor]. If this anchor was persistent you must first call [method make_anchor_unpersistent] and await its callback.
```
> If this anchor was persistent you must first call [method make_anchor_unpersistent] and await its callback.

Isn't something we could do automatically?

BastiaanOlij replied:
I thought about that but it has problems because making an anchor no longer persistent, is an async method. Especially if removing anchors is part of some cleanup logic, we potentially need to ensure the user delays part of their cleanup.

I'm on the fence though. We could make removing an anchor an asynchronous method on our side, even if it would be immediate if the anchor is not persistent and deal with it that way.

m4gr3d commented on modules/openxr/doc_classes/OpenXRSpatialAnchorCapability.xml:
```diff
@@ -0,0 +1,120 @@
+		</method>
+	</methods>
+	<signals>
+		<signal name="anchor_persistent_failed">
```
`persistent` should be a name here, i.e: `anchor_persistence_failed`.

m4gr3d commented on modules/openxr/extensions/spatial_entities/openxr_spatial_anchor.cpp:
```diff
@@ -0,0 +1,1206 @@
+}
+
+////////////////////////////////////////////////////////////////////////////
+// OpenXRSpatialContextPersistenceConfig1
```
```suggestion
// OpenXRSpatialContextPersistenceConfig
```

m4gr3d commented on modules/openxr/extensions/spatial_entities/openxr_spatial_anchor.cpp:
```diff
@@ -0,0 +1,1206 @@
+HashMap<String, bool *> OpenXRSpatialAnchorCapability::get_requested_extensions() {
+	HashMap<String, bool *> request_extensions;
+
+	if (GLOBAL_GET("xr/openxr/extensions/spatial_entity/enabled") && GLOBAL_GET("xr/openxr/extensions/spatial_entity/enable_spatial_anchors")) {
+		request_extensions[XR_EXT_SPATIAL_ANCHOR_EXTENSION_NAME] = &spatial_anchor_ext;
+		if (GLOBAL_GET("xr/openxr/extensions/spatial_entity/enable_persistent_anchors")) {
```
Should update instead of `GLOBAL_GET` with `GLOBAL_GET_CACHED` instead.

BastiaanOlij replied:
We'll need to do a PR to clean this up in older extensions as well.

m4gr3d commented on modules/openxr/extensions/spatial_entities/openxr_spatial_anchor.cpp:
```diff
@@ -0,0 +1,1206 @@
+
+	se_extension->connect(SNAME("spatial_discovery_recommended"), callable_mp(this, &OpenXRSpatialAnchorCapability::_on_spatial_discovery_recommended));
+
+	if (GLOBAL_GET("xr/openxr/extensions/spatial_entity/enable_builtin_anchor_detection")) {
```
Similar comment about switching to `GLOBAL_GET_CACHED`

m4gr3d commented on modules/openxr/extensions/spatial_entities/openxr_spatial_anchor.cpp:
```diff
@@ -0,0 +1,1206 @@
+bool OpenXRSpatialAnchorCapability::_load_supported_persistence_scopes() {
+	ERR_FAIL_COND_V(!spatial_persistence_ext, false);
+
+	if (supported_persistence_scopes.size() == 0) {
```
```suggestion
	if (supported_persistence_scopes.is_empty()) {
```

m4gr3d commented on modules/openxr/extensions/spatial_entities/openxr_spatial_anchor.cpp:
```diff
@@ -0,0 +1,1206 @@
+		}
+
+		if (is_print_verbose_enabled()) {
+			if (supported_persistence_scopes.size() > 0) {
```
```suggestion
			if (!supported_persistence_scopes.is_empty()) {
```

m4gr3d commented on modules/openxr/extensions/spatial_entities/openxr_spatial_plane_tracking.cpp:
```diff
@@ -0,0 +1,713 @@
+HashMap<String, bool *> OpenXRSpatialPlaneTrackingCapability::get_requested_extensions() {
+	HashMap<String, bool *> request_extensions;
+
+	if (GLOBAL_GET("xr/openxr/extensions/spatial_entity/enabled") && GLOBAL_GET("xr/openxr/extensions/spatial_entity/enable_plane_tracking")) {
```
Similar comment about using `GLOBAL_GET_CACHED` instead.

BastiaanOlij commented on modules/openxr/extensions/spatial_entities/openxr_spatial_entity_extension.h:
```diff
@@ -0,0 +1,213 @@
+		OPENXR_SPATIAL_COMPONENT_TYPE_PERSISTENCE = 1000763000,
+	};
+
+	// TODO Q maybe register all spatial entity enums here,
+	// currently we define a few in various components and capabilities.
```
I'm still unsure about this. The logical place would be to add these to `OpenXRInterface` or `OpenXRSpatialEntityExtension`, but most places that need to use these constants, will be defined before we define these classes. So we have a dependency issue.

That said, if I can expose the OpenXR enums directly to ClassDB instead of creating mirror enums, this may become a none issue.

BastiaanOlij commented on modules/openxr/doc_classes/OpenXRSpatialCapabilityConfigurationPlaneTracking.xml:
```diff
@@ -0,0 +1,38 @@
+				[b]Note:[/b] Only valid after this configuration was used to create a spatial context.
+			</description>
+		</method>
+		<method name="get_supports_labels">
+			<return type="bool" />
+			<description>
+				Returns [code]true[/code] if we support the plane semantic label component (only valid after the OpenXR session has started). You can query these using the [OpenXRSpatialComponentPlaneSemanticLabelList] data object.
+			</description>
+		</method>
+		<method name="get_supports_mesh_2d">
+			<return type="bool" />
+			<description>
+				Returns [code]true[/code] if we support the mesh 2d component (only valid after the OpenXR session has started). You can query these using the [OpenXRSpatialComponentMesh2DList] data object.
+			</description>
+		</method>
+		<method name="get_supports_polygons">
+			<return type="bool" />
+			<description>
+				Returns [code]true[/code] if we support the polygon 2d component (only valid after the OpenXR session has started). You can query these using the [OpenXRSpatialComponentPolygon2DList] data object.
+			</description>
+		</method>
```
Just habit when defining getters and setters, I'll change them :)

BastiaanOlij replied:
Uhm, and the name prevents issues with the variable being called the same as the method, hmzl..

BastiaanOlij commented on modules/openxr/extensions/spatial_entities/openxr_spatial_anchor.h:
```diff
@@ -0,0 +1,251 @@
+	GDCLASS(OpenXRSpatialAnchorCapability, OpenXRExtensionWrapper);
+
+public:
+	enum OpenXrSpatialPersistenceScope { // Should mirror XrSpatialPersistenceScopeEXT
```
I'm now using `XrSpatialPersistenceScopeEXT` directly, but still exposing it on `OpenXRSpatialAnchorCapability`.

BastiaanOlij commented on modules/openxr/openxr_structure.h:
```diff
@@ -0,0 +1,72 @@
+	virtual void *get_header(void *p_next = nullptr);
+
+	void set_next(const Ref<OpenXRStructureBase> p_next);
+	Ref<OpenXRStructureBase> get_next() const;
```
Ok, we're now exposing `XrStructureType` as an enum :) No need for int. I didn't go through the pain of adding descriptions to every entry though, there are ALOT.

BastiaanOlij commented on modules/openxr/extensions/spatial_entities/openxr_spatial_entity_extension.h:
```diff
@@ -0,0 +1,213 @@
+public:
+	// Enums that mirror important relevant OpenXR enums (should see if we can macro this)
+
+	enum OpenXrSpatialComponentType { // Must mirror XrSpatialComponentTypeEXT
```
And here :)

BastiaanOlij replied:
This has now changed to `ComponentType`, also exposed `Capability` as an enum. They now shadow their OpenXR counterparts.

BastiaanOlij commented on modules/openxr/extensions/spatial_entities/openxr_spatial_marker_tracking.h:
```diff
@@ -0,0 +1,264 @@
+	GDCLASS(OpenXRSpatialCapabilityConfigurationAprilTag, OpenXRSpatialCapabilityConfigurationBaseHeader);
+
+public:
+	enum OpenXrSpatialMarkerAprilTagDict { // Mirror XrSpatialMarkerAprilTagDictEXT
```
And here :)

BastiaanOlij replied:
This has now changed to AprilTagDict , also exposed AprilTagDict as an enum. They now shadow their OpenXR counterparts.

BastiaanOlij commented on modules/openxr/extensions/spatial_entities/openxr_spatial_plane_tracking.h:
```diff
@@ -0,0 +1,226 @@
+	// TODO add enum that mirrors XrSpatialPlaneAlignmentEXT
+
+public:
+	enum OpenXrSpatialPlaneAlignment { // Must mirror XrSpatialPlaneAlignmentEXT
```
And here :)

BastiaanOlij commented:
@dsnopek @m4gr3d 

Ok, I've worked through most of the feedback. Few questions and remarks I've left open up above, so feel free to mark those as resolved if you're happy with the feedback.

The biggest structural change is that I've added some XR versions of our binding macros that allow us to bind OpenXR enums and expose them in Godot. I'm pretty happy with this approach and we should slowly clean up existing implementations to do the same.
The only discussion left here is whether we're happy to have the enums live on the objects they most directly related to, or if we should introduce a container class or something. It is possible that the issues here I originally ran into are already solved with improvements I made to the macros later. I'll do some more testing with that tomorrow.

The only outstanding task is to complete the logic for the meshes/collision shapes on Mesh2D and Polygon2D entities coming from plane tracking. Should be able to do that over the weekend.

BastiaanOlij commented:
Marked all classes introduced as experimental.

dsnopek commented on modules/openxr/doc_classes/OpenXRSpatialAnchorCapability.xml:
```diff
@@ -0,0 +1,120 @@
+				[b]Note:[/b] This is an asynchronous method. This method returns false if we fail to start the method. On success [param callback] will be called. On failure [signal anchor_persistent_failed] will be emitted.
+			</description>
+		</method>
+		<method name="make_anchor_unpersistent">
```
> This is the naming used in OpenXR (e.g. `xrUnpersistSpatialEntityAsyncEXT`) but I agree it feels wrong.

I think "unpersist" is a word, though. You can unpersist something, ie removing its persistent property.

And something can be "not persistent" (to constrast it from persistent things). There is also "non-persistent" which means temporary (or, actually, more temporary than normal, because otherwise you wouldn't say anything if it was just normal). But I've never personally heard of something being "unpersistent"

BastiaanOlij replied:
As discussed on rocket chat, we settled on `persist_anchor()` and `unpersist_anchor()`, this has now been changed.

dsnopek commented:
> The biggest structural change is that I've added some XR versions of our binding macros that allow us to bind OpenXR enums and expose them in Godot. I'm pretty happy with this approach and we should slowly clean up existing enums to do the same.

I think directly exposing OpenXR enums can make sense in the situations were we are effectively directly exposing the OpenXR API, for example, via `OpenXRStructureBase` and its children. Or, even on the extension wrapper classes, when those are meant to be the lower-level interface.

However, for anything that the average developer who just wants to use spatial entity functionality has to interact with, I think this will be confusing. Especially once all the vendor extensions start appearing.

For example, right now we've got:

- `XR_SPATIAL_PLANE_SEMANTIC_LABEL_WALL_EXT`
- `XR_SPATIAL_PLANE_SEMANTIC_LABEL_TABLE_EXT`

... but this list could get expanded when vendor extensions add (hypothetical) items like:

- `XR_SPATIAL_PLANE_SEMANTIC_LABEL_SCREEN_META`
- `XR_SPATIAL_PLANE_SEMANTIC_LABEL_WINDOW_BD`
- `XR_SPATIAL_PLANE_SEMANTIC_LABEL_SHELF_ANDROID`

I think for higher-level stuff, our tendency is to hide these sort of things behind friendlier interfaces. Although, figuring out where to draw the line between low-level and high-level can be tricky too.

So, yeah, something we will need to continue to discuss! Probably would be easiest in a call or at an XR team meeting

dsnopek commented on modules/openxr/doc_classes/OpenXRStructureBase.xml:
```diff
@@ -0,0 +1,1068 @@
+		</member>
+	</members>
+	<constants>
+		<constant name="XR_TYPE_UNKNOWN" value="0" enum="XrStructureType">
```
Given that new values are added to this enum in almost every release of OpenXR (since this includes all the values from all extensions too), documenting these values in any way (even just linking to the spec) might be a sisyphean task.

Perhaps this one shouldn't be exposed as an enum, and we should just use an integer? Anyone who would directly interact with this value has to have the OpenXR spec open while they're working anyway :-)

BastiaanOlij replied:
Indeed, I think this one is really difficult, especially seeing we haven't used this approach for many existing APIs. I think as we make the system more extendable, we'll make more use of it.

Whether it will start to make more sense to expose this enum in its entirety, or whether we should hide this behind an int, idk. Both are defensible.

dsnopek replied:
Well, if the idea is that this system should be extendable from GDExtension, we don't need to expose the enum, because from GDExtension you can use the OpenXR headers and those enums directly. GDExtension only needs the integer.

We really only need to expose the enum if we think this will be used by GDScript. Which I don't think will be the case? (Except maybe if someone working on a GDExtension is doing some debugging/testing using GDScript?)

BastiaanOlij replied:
As discussed during the XR meeting awhile back, we've removed the `XrStructureType` exposure and are now exposing an integer to GDScript/GDExtension.

BastiaanOlij commented:
Eventually that will mean adding additional enum values in, potentially through GDExtensions. 

The reason I'm exposing this level of detail is that those who wish to build those extensions, will need that exposed. It's why I'm exposing even these core objects. Normal users won't ever use them, but those who want to go beyond the base logic will.

This is one thing we haven't really got something for, how do we expose something that is not meant for the average user, only for those who are extending the system?

BastiaanOlij commented:
> I think normal users will want semantic labels, for example, though. And right now this is being exposed as the OpenXR enum directly.

Owh right, I had forgotten I had also exposed this as is on the `OpenXRPlaneTracker` tracker object, I was thinking of the `OpenXRSpatialComponentPlaneSemanticLabelList` which normal users won't use.

I think we're kinda stuck here whatever we use our own enum, or that of OpenXR, we don't really have a solid way of adding more enum values to an existing enum. This will always require an update of the thirdparty OpenXR files.

This is one of the areas where I'm not sure how vendor extensions will be able to modify core logic and I think that is a bridge we should cross once those extensions are introduced and we can see how they work from an OpenXR POV and if there are changes needed here.

BastiaanOlij commented:
Rebased, cleaned up a few more things, and implement the missing bits of the `get_mesh` and `get_shape` implementation for `OpenXRPlaneTracker`.

Still loads of things to further discuss but I think implementation wise everything is complete now.

Only real core issue preventing merging as experimental is the discussion David raised, while all the "low level" implementations expose things as one would expect, some of this shining through to specifically the implementation of `OpenXRPlaneTracker`, may be something we want to do differently. How however is currently up for debate as we do not yet know how this will evolve in OpenXR land and whether this poses a real problem or not.
There is an argument that us marking this class as experimental will allow us to tweak the API if we do want to change this part later on.

The decision IMHO here will depends on whether this remains the only real issue that needs further discussion, or if other shortcomings are found that make us unsure of the implementation.

BastiaanOlij commented:
Decisions in XR Meeting:
1. remove XrStructureType enum, overkill
2. change semantic label on the Tracker to a string, then evaluate if it still makes sense to have the enum in the `OpenXRSpatialComponentPlaneSemanticLabelList` class (or just expose an untyped integer).
3. evaluate the others on a case by case basis. Definite bias to keeping the mirrored approach instead of exposing OpenXR enums directly.

BastiaanOlij commented:
Ok just pushed a number of changes, I'm not at the end yet and need to clean up a bit of testing code but made the following changes:
1. As discussed `XrStructureType` has been removed as an enum. We're now exposing an int to GDScript/GDExtension
2. The semantic label changes have been done. We now have an enum that mirrors the OpenXR one but is only exposed on the internal `OpenXRSpatialComponentPlaneSemanticLabelList` object. `OpenXRPlaneTracker` now exposes a string.
3. I'm now lazy loading capabilities, we were doing this when the OpenXR instance was created but this was too early, the data is not available until the session was created. As we can't guarantee order of execution between extensions, I'm lazy loading it.
4. There is now a throttle on how often we check plane tracking. Runtimes are supposed to throttle this but we have no guarantee here. We now set a flag when the signal for new plane tracking data is received from the runtime and wait with reacting until 60 frames have passed since the last update. 60 is an arbitrary choice atm and we may wish to make this a setting.
5. We're now checking if the plane tracking data has actually changed with a bit of a margin (> 1 mm) and thus not updating geometry unless needed

Still todo:
- Expose `XrSpatialPlaneAlignmentEXT` as a mirrored enum (and probably check others)
- Test and bugfix markers

BastiaanOlij commented on modules/openxr/extensions/spatial_entities/openxr_spatial_plane_tracking.h:
```diff
@@ -0,0 +1,226 @@
+	GDCLASS(OpenXRSpatialComponentPlaneSemanticLabelList, OpenXRSpatialComponentData);
+
+public:
+	enum OpenXrSpatialPlaneSemanticLabel { // Must mirror XrSpatialPlaneSemanticLabelEXT
```
As discussed during the XR meeting a week or two back, we've decided to go back to our own struct with naming as suggested by David.
I'm assigning the OpenXR constants to our constants to make extra clear we're duplicating those for our internal needs.

BastiaanOlij commented on modules/openxr/extensions/spatial_entities/openxr_spatial_marker_tracking.h:
```diff
@@ -0,0 +1,264 @@
+	GDCLASS(OpenXRSpatialCapabilityConfigurationAruco, OpenXRSpatialCapabilityConfigurationBaseHeader);
+
+public:
+	enum OpenXrSpatialMarkerArucoDict { // mirror XrSpatialMarkerArucoDictEXT
```
This has now changed to ArucoDict , also exposed ArucoDict as an enum. They now shadow their OpenXR counterparts.

BastiaanOlij commented on modules/openxr/extensions/spatial_entities/openxr_spatial_entities.cpp:
```diff
@@ -0,0 +1,505 @@
+	OpenXRSpatialEntityExtension *se_extension = OpenXRSpatialEntityExtension::get_singleton();
+	ERR_FAIL_NULL_V(se_extension, PackedVector2Array());
+
+	return se_extension->get_vector2_buffer(p_snapshot, buffer.bufferId);
```
Just as a reminder to myself, I'll probably do this in a follow up.
Currently we retrieve this data and check in `set_mesh_data` if it has changed, which we should keep as a fallback.

However feedback from the OpenXR WG suggests that we can add an extra check here. We should cache `buffer.bufferId` in our calling logic, if it is unchanged from the last time, the data is unchanged and we don't need to call this getter.

When new data is available, we should always have a new buffer id.

This applies to our other buffer getter functions as well.

BastiaanOlij commented:
XR Meeting: this is ready to merge, if there are any more tweaks we can PR them separately but everything seems to be working well.

dsnopek reviewed (approved):
Thanks!

I've done one last round of testing, and everything seems to be working well. I have a couple notes that I think should get handled, but they can come in follow-up PRs

BastiaanOlij commented on modules/openxr/extensions/spatial_entities/openxr_spatial_anchor.cpp:
```diff
@@ -0,0 +1,1206 @@
+	}
+
+	// Create our future result
+	Ref<OpenXRFutureResult> future_result = future_api->register_future(future, callable_mp(this, &OpenXRSpatialAnchorCapability::_on_persistence_context_ready).bind(p_user_callback).bind((uint64_t)p_scope));
```
Dang, I didn't know that was possible, learn something every day:)

BastiaanOlij replied:
Before I introduce a bug, when adding them into a single bind, does the order still need to be reversed?

BastiaanOlij commented on modules/openxr/doc_classes/OpenXRSpatialAnchorCapability.xml:
```diff
@@ -0,0 +1,123 @@
+		<method name="is_spatial_anchor_supported">
+			<return type="bool" />
+			<description>
+				Returns [code]true[/code] if spatial anchors is supported by the hardware. Only returns a valid value after OpenXR is initialized.
```
I think that's `Only returns a valid value after OpenXR has been initialized.` ?

BastiaanOlij commented on doc/classes/ProjectSettings.xml:
```diff
@@ -3449,6 +3449,38 @@
+		</member>
+		<member name="xr/openxr/extensions/spatial_entity/enabled" type="bool" setter="" getter="" default="false">
+			If [code]true[/code], support for the spatial entity extension is requested. If supported, you will be able to access spatial information about the real environment around you. What information is dependent on additional extensions.
+			[b]Note:[/b] This requires the OpenXR data source extension and unobstructed handtracking to be supported by the XR runtime. If not supported this setting will be ignored.
```
Oops, copy/paste and forgetting to adjust it ;)

BastiaanOlij commented:
@AThousandShips I did most of your clean up suggestions, there are one or two outstanding questions and I didn't do all the changes away from `nullptr`, too easy to mess up.

I suggest that whatever is left we do as a separate PR, that was a LOT to get through :)

AThousandShips commented on modules/openxr/extensions/spatial_entities/openxr_spatial_marker_tracking.cpp:
```diff
@@ -0,0 +1,698 @@
+
+Ref<OpenXRFutureResult> OpenXRSpatialMarkerTrackingCapability::_create_spatial_context() {
+	OpenXRSpatialEntityExtension *se_extension = OpenXRSpatialEntityExtension::get_singleton();
+	ERR_FAIL_NULL_V(se_extension, nullptr);
```
It's explicit in a long method, helps readability

Repiteo commented:
Thanks! Great work covering all those bases!
